### ISSUE TITLE
[$1000] WS - Screen jerks when opening member page

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#27271]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open any WS
2. Click member

## Expected Result:
The screen is not jerky

## Actual Result:
The screen jerks

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** v1.3.67-0

**Reproducible in staging?:** Y

**Reproducible in production?:** Y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/115492554/af91797c-1049-4b76-a904-71a6c3dfa434

https://github.com/Expensify/App/assets/115492554/648bfa66-5ca1-441b-9962-6d5802373f0e

**Expensify/Expensify Issue URL:**

**Issue reported by:** @namhihi237

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1694195603643719

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01883954f0a88ab38c</li>
        <li>Upwork Job ID: 1701639747168940032</li>
        <li>Last Price Increase: 2023-10-07</li>
    </ul>
</details>

**Comment by b4s36t4** [comment]
This needs Hold for `SelectionList` refactor PR, because that solves the issue (not entirely sure).

**Comment by studentofcoding** [comment]
IMO, the jerking is very subtle (or more like a small delay), and not really bother the UX

**Comment by mallenexpensify** [comment]
> This needs Hold for SelectionList refactor PR, because that solves the issue (not entirely sure).
@ArekChr do you agree? If so.. I see the below as the tracking list, is there a specific issue/PR this should be helpd on?
- https://github.com/Expensify/App/issues/11795

@MitchExpensify , unassigning since you were assigned first.

**Comment by MitchExpensify** [comment]
It is subtle but worth fixing IMO, waiting on proposals here  Melvin

**Comment by MitchExpensify** [comment]
Waiting for proposals

**Comment by MitchExpensify** [comment]
Same

**Comment by MitchExpensify** [comment]
Doubling to get some more proposals

**Comment by b4s36t4** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
WS - Screen jerks when opening member page
### What is the root cause of that problem?

We have autofocus setup for the input inside members page which is causing the issue. If we observe the focus effect in the attached video from description input is getting focused while the transition is still going (250 milliseconds delay so the reason of subtle jerk). 

The amount of delay we had (300ms) is not entirely working perfect. Depends on device, speed and the amount of calculations we do the autofocus works.

To keep it simple, While the screen transition is going, the input is getting focused which is causing the issue while the the issue is so subtle but it happens everytime if the screen is new (opening after refresh or pop the screen and open it again).

While the `useFocusEffect` approach is solving the issue but it's not entirely 100% good approach. `useFocusEffect` get's called immediately when screen starts navigation and after 300ms the input is getting focused in which my view is a work around.

### What changes do you think we should make in order to solve the problem?

We have `transitionEnd` event subscribed for navigation inside `ScreenWrapper` which is working perfect in most cases whereas it's not (around 50% of time) not getting fired because of the `Onyx` calculations and the event expiration on refresh. If I do a refresh on members page input won't get focused most of time.

To fix this we should rely on `transitionEnd` event only which fires the event when screen is fully completed navigation so that there shouldn't be any jerk or breaking.

#### Solution

Introducing a hook to observe `transitionEnd` event, inspired form `useFocusEffect`.

```jsx
import {useNavigation} from '@react-navigation/native';
import {useEffect, useState} from 'react';

export default (callback) => {
    const navigation = useNavigation();

    const [isCallbackInitiated, setIsCallbackInitiated] = useState(false);

    useEffect(() => {
        // If the screen is focused & callback is already done, don't run the event again
        if (navigation.isFocused() && isCallbackInitiated) {
            return;
        }
        if (isCallbackInitiated) {
            return;
        }
        const unsubscribe = navigation.addListener('transitionEnd', () => {
            setIsCallbackInitiated(true);
            callback();
        });
        return unsubscribe;
    }, [navigation, callback, isCallbackInitiated]);
};

```

> Tried running it on all platforms working seamlessly, although this might can cause few issues (need help one few) we might need QA support.




### What alternative solutions did you explore? (Optional)
Increase the timeout for native devices like for 600MS or so?

**Comment by MitchExpensify** [comment]
@ArekChr proposal ready for your review

**Comment by MitchExpensify** [comment]
Friendly bump @ArekChr

**Comment by ArekChr** [comment]
@b4s36t4 Using `transitionEnd` for focusing the text input won’t work when popping a screen off the stack. This method doesn’t fire when returning from the "Get assistance" screen (? button) to Members

**Comment by b4s36t4** [comment]
So increasing the delay for native devices would be the way?! @ArekChr

**Comment by ArekChr** [comment]
@b4s36t4 Yes, but still considering if it's worth increasing the delay on Android as well, as the transition time is shorter for this platform, I am also looking for more alternative solutions

**Comment by MitchExpensify** [comment]
What do we need to keep this moving @ArekChr ?

**Comment by MitchExpensify** [comment]
Friendly bump @b4s36t4

**Comment by b4s36t4** [comment]
Getting back at this today.

Thanks @MitchExpensify for the ping :) totally forget!!

**Comment by MitchExpensify** [comment]
Great, no worries @b4s36t4 !

**Comment by MitchExpensify** [comment]
What are your next steps here @b4s36t4 ?

**Comment by b4s36t4** [comment]
I'm still trying to debug more & find a fix for the issue. Probably expecting a solution in next 2 days.

**Comment by MitchExpensify** [comment]
Ok sounds good, looking forward to hearing how that goes!

**Comment by MitchExpensify** [comment]
Any update @b4s36t4 ?

**Comment by MitchExpensify** [comment]
bump @b4s36t4, thanks!

**Comment by MitchExpensify** [comment]
This issue is starting to drag on quite a bit, any reliable estimate on when you can take a look @b4s36t4 ? I'm considering reassigning if this doesn't get attention soon

**Comment by MitchExpensify** [comment]
No bites yet

**Comment by keisyrzk** [comment]
Hello! I'm Krzysztof from Callstack - an expert contributor group and I would like to work on this one.

**Comment by keisyrzk** [comment]
FYI, I do have partial solutions for the moment but it is more performance issue where two animations (text input focus & navigation transition) want to run at the same time and it blocks the thread. Investigating and working on a better solution.

**Comment by MitchExpensify** [comment]
How is the better solution coming along, @keisyrzk ?

**Comment by keisyrzk** [comment]
@MitchExpensify Unfortunately no better solution so far :/ 
For the moment I can see two temporary solutions:

1. Increasing the animation delay time from 300ms to minimum 500ms as this value seems to be enough as the transition animation last about 400ms.
Unfortunately if anything happens that the transition animation last longer, the "issue" returns. The pros here is a change in one global place so the update would affect whole app at once.

2. The second idea is based on a state management. We could try to use `transitionEnd` event and then focus the text field. The thing here is that it works if you come to the view with transition (so the transitionEnd event indeed happens). Unfortunately if you reresh the view it will not work as there is no `transition` and no `transitionEnd` event. We could try to implement a check if the "transition is going to happen" but it means more complicated code, more possibilities for paths that may cause other issues and it would be a view specific solution so to deal with the issue globally we would need to implement it everywhere separately or create some kind of a manager.

As you may see both possibilities are not something we would like to implement. I asked the performance team if they have some other ideas.

**Comment by keisyrzk** [comment]
PROPOSAL

**Please re-state the problem that we are trying to solve in this issue.**
When loading the “Members” view from “Workspace” the “Members” view is lagging on appearing

**What is the root cause of that problem?**
The root of the problem states for transition between views and TextInput focus animations overlapping on the same thread. Currently TextInput focusing is delayed by the 300ms to satisfy the animations queue but it is not enough as the transition last about 420ms and may vary over time and in different circumstances.

**What changes do you think we should make in order to solve the problem?**
The issue might be fixed by modifying WorkspaceMembersPage function as follows:
`    const textInputRef = useRef(null);
    const [hasTransitionEnded, setHasTransitionEnded] = useState(false);

    const onTransitionEnded = () => {
        setHasTransitionEnded(true);
        textInputRef.current.focus();
    }

    // Effect for component mount and update
    useEffect(() => {
        // If the component is rendered without a transition, focus immediately
        if (!hasTransitionEnded) {
            textInputRef.current.focus();
        }
    }, [hasTransitionEnded]);`

and updating the ScreenWrapper's onEntryTransitionEnd prop like so:
`    return (
        <ScreenWrapper
            includeSafeAreaPaddingBottom={false}
            style={[styles.defaultModalContainer]}
            testID={WorkspaceMembersPage.displayName}
            onEntryTransitionEnd={onTransitionEnded}
(...)`

**What alternative solutions did you explore? (Optional)**
None worth applying on production.

**Comparison measurements showing the performance result of the proposal**
![Zrzut ekranu 2023-12-7 o 09 44 04](https://github.com/Expensify/App/assets/18525360/313ebfc9-8147-4b03-8f71-4937eff3e67b)

**Comment by b4s36t4** [comment]
What if we just reduce the animation time for this particular page? I think with react-navigation we can tweak the animation time for the screen, doing so would also work in my view. to the deeper question, why does this page only getting delayed with animation (as per the report above) doesn't we use same methods everywhere?

**Comment by keisyrzk** [comment]
This is a problem present in whole app. Manipulating the `time / delays` does not guarantee us stable result.

**Comment by b4s36t4** [comment]
But using `onEntryTransitionEnd` doesn't work. more on this https://github.com/Expensify/App/issues/27271#issuecomment-1746354259.

**Comment by b4s36t4** [comment]
also this `onEntryTransitionEnd` doesn't fire on first instance of opening app, the main reason we switched to `useFocusEffect` method.

**Comment by keisyrzk** [comment]
we can go below way of the `onEntryTransitionEnd` does not work for you

```js
const navigation = useNavigation();
    const textInputRef = useRef(null);
    const [hasTransitionEnded, setHasTransitionEnded] = useState(false);

    // Listener for the end of transition
    useEffect(() => {
        const unsubscribe = navigation.addListener('transitionEnd', () => {
            setHasTransitionEnded(true);
            textInputRef.current.focus();
        });

        return unsubscribe;
    }, [navigation]);
```

**Comment by b4s36t4** [comment]
`transitionEnd` event doesn't fire when we refresh or poping up a screen, which is a problem we don't need.

**Comment by keisyrzk** [comment]
yes indeed, this is why we do also have

```js
useEffect(() => {
        // If the component is rendered without a transition, focus immediately
        if (!hasTransitionEnded) {
            textInputRef.current.focus();
        }
    }, [hasTransitionEnded]);
```

if you test my proposal you will see that it is working on refresh as well. I do also have videos already recorded. I can share them if this help.

**Comment by MitchExpensify** [comment]
@b4s36t4 What is your take on @keisyrzk 's proposal at this point?

**Comment by ArekChr** [comment]
Hello @MitchExpensify, I'm unassigning myself - I left the c+ team and joined Waves.

**Comment by situchan** [comment]
@MitchExpensify I am interested in reviewing this

**Comment by MitchExpensify** [comment]
Assigned to you @situchan, lets keep this moving 🙇

**Comment by MitchExpensify** [comment]
Friendly bump @situchan

**Comment by situchan** [comment]
This is tricky issue.
@MitchExpensify before researching more, can you please confirm if this fits to any part of waves?

**Comment by MitchExpensify** [comment]
No in hindsight I do not think it qualifies for a wave
